lib/linked_list.rb

```diff
@@ -18,128 +18,237 @@ def initialize
     end
 
 
-    # Additional Exercises 
+    # Additional Exercises
     # returns the value in the first node
     # returns nil if the list is empty
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def get_first
-      raise NotImplementedError
+      @head ? @head.data : nil
     end
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def add_last(value)
-      raise NotImplementedError
+      new_node = Node.new(value)
+      if @head.nil?
+        @head = new_node
+      else
+        current_node = @head
+        until current_node.next.nil?
+          current_node = current_node.next
+        end
+        current_node.next = new_node
+      end
     end
 
     # method that returns the length of the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def length
-      raise NotImplementedError
+      length = 0
+      current_node = @head
+      while current_node
+        length += 1
+        current_node = current_node.next
+      end
+      return length
     end
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns nil if there are fewer nodes in the linked list than the index value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def get_at_index(index)
-      raise NotImplementedError
+      return nil if index > self.length || self.length == 0
+      node_index = 0
+      current_node = @head
+      until node_index == index
+        current_node = current_node.next
+        node_index += 1
+      end
+      return current_node.data
     end
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def visit
-      raise NotImplementedError
+      current_node = @head
+      while current_node
+        print current_node.data
+        print ' '
+        current_node = current_node.next
+      end
     end
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def delete(value)
-      raise NotImplementedError
+      return nil if @head.nil?
+      current_node = @head
+      previous_node = nil
+      if current_node.data == value
+        @head = current_node.next
+      else
+        until current_node.data == value
+          previous_node = current_node
+          current_node = current_node.next
+        end
+        if current_node.next.nil?
+          previous_node.next = nil
+        else
+          previous_node.next = current_node.next
+        end

```

Comment:
Do you need this if/else or can you condense this into:

```suggestion
        previous_node.next = current_node.next


```